packages/runtime/src/runtime.ts

```diff
@@ -324,19 +347,160 @@ export class Runtime {
+      throw new Error("Only Freeze account can freeze asset");
     }
+  }
 
-    const creatorAcc = this.assertAccountDefined(addr, this.store.accounts.get(addr));
-    const assetDef = creatorAcc.getAssetDef(assetId);
-    return this.assertAssetDefined(assetId, assetDef);
+  /**
+   * https://developer.algorand.org/docs/features/asa/#freezing-an-asset
+   * Freezes assets for a target account
+   * @param sender sender address
+   * @param assetId asset index
+   * @param freezeTarget target account
+   * @param freezeState target state
+   * @param payFlags transaction parameters
+   */
+  freezeAsset (
+    sender: string, assetId: number, freezeTarget: string,
+    freezeState: boolean, payFlags: TxParams
+  ): void {
+    this.mkAssetFreezeTx(sender, assetId, freezeTarget, freezeState, payFlags);
+    const acc = this.assertAccountDefined(freezeTarget, this.store.accounts.get(freezeTarget));
+    acc.freezeAsset(assetId, freezeState);
+  }
+
+  // creates asset revoke transaction object.
+  mkAssetRevokeTx (
+    sender: string, recipient: string, assetId: number,
+    revocationTarget: string, amount: number, flags: TxParams, assetID: number
+  ): void {
+    // this funtion is called only for validation of parameters passed
+    algosdk.makeAssetTransferTxnWithSuggestedParams(
+      sender,
+      recipient,
+      flags.closeRemainderTo,
+      revocationTarget,
+      amount,
+      encodeNote(flags.note, flags.noteb64),
+      assetId,
+      mockSuggestedParams(flags, this.round)
+    );
+    const asset = this.getAssetDef(assetID);
+    if (asset.clawback !== sender) {
+      throw new Error("Only Clawback account can revoke assets");
+    }
+  }
+
+  /**
+   * https://developer.algorand.org/docs/features/asa/#revoking-an-asset
+   * Revoking an asset for an account removes a specific number of the asset
+   * from the revoke target account.
+   * @param sender sender address
+   * @param recipient asset receiver address
+   * @param assetId asset index
+   * @param revocationTarget revoke target account
+   * @param amount amount of assets
+   * @param payFlags transaction parameters
+   */
+  revokeAsset (
+    sender: string, recipient: string,
+    assetID: number, revocationTarget: string,
+    amount: number, payFlags: TxParams
+  ): void {
+    this.mkAssetRevokeTx(sender, recipient, assetID, revocationTarget, amount, payFlags, assetID);
+    // Transfer assets
+    const fromAssetHolding = this.getAssetHolding(assetID, revocationTarget);
+    const toAssetHolding = this.getAssetHolding(assetID, recipient);
+    this.assertAssetNotFrozen(assetID, revocationTarget);
+    this.assertAssetNotFrozen(assetID, recipient);
+
+    if (fromAssetHolding.amount - amount < 0) {
+      throw new TealError(ERRORS.TRANSACTION.INSUFFICIENT_ACCOUNT_ASSETS, {
+        amount: amount,
+        address: revocationTarget
+      });
+    }
+    fromAssetHolding.amount -= amount;
+    toAssetHolding.amount += amount;
```

Comment:
Can we use the `runtime.transferAsset` function here?

Reply:
no,If you look at code we need secret key for the revocationTarget and we don't have that.